File: 560-SubarraySumEqualsK/560-SubarraySumEqualsK.java
/*

560. Subarray Sum Equals K
Medium

Given an array of integers and an integer k, you need to find the total number of continuous subarrays whose sum equals to k.

Example 1:
Input:nums = [1,1,1], k = 2
Output: 2
Note:
The length of the array is in range [1, 20,000].
The range of numbers in the array is [-1000, 1000] and the range of the integer k is [-1e7, 1e7].

*/

class Solution {
    public int subarraySum(int[] nums, int k) {
        //brute force
        int res = 0;
        for(int i =0; i < nums.length; i++){
            int count = 0;
            for(int j = i; j < nums.length; j++ ){
                count += nums[j];
                if (count == k){
                    res++;
                }
            }
        }
        return res;
    }
}

File: 560-SubarraySumEqualsK/560-SubarraySumEqualsK.py
"""

560. Subarray Sum Equals K

Given an array of integers nums and an integer k, return the total number of subarrays whose sum equals to k.

A subarray is a contiguous non-empty sequence of elements within an array.

 

Example 1:

Input: nums = [1,1,1], k = 2
Output: 2

Example 2:

Input: nums = [1,2,3], k = 3
Output: 2

 

Constraints:

    1 <= nums.length <= 2 * 104
    -1000 <= nums[i] <= 1000
    -107 <= k <= 107

"""

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        mapp = {0: 1}
        res, summ = 0, 0
        for n in nums:
            summ += n
            if (summ - k) in mapp:
                res += mapp.get(summ - k)
            mapp[summ] = mapp.get(summ, 0) + 1
        return res
